File: not-proven/proven-deception/ffi/zig/src/deception.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// deception.zig -- Zig FFI implementation of proven-deception.
//
// Implements the deception/decoy server state machine with:
//   - 64-slot mutex-protected server pool
//   - Decoy deployment per server (max 32 decoys)
//   - Alert tracking per server (max 64 alerts)
//   - Trigger event matching and alert priority assignment
//   - Response action execution per triggered decoy
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching DeceptionABI.Types.idr exactly.

const std = @import("std");

// =========================================================================
// Enums (matching DeceptionABI.Types.idr tag assignments)
// =========================================================================

/// Decoy asset types (ABI tags 0-5).
pub const DecoyType = enum(u8) {
    service = 0,
    credential = 1,
    file = 2,
    network = 3,
    token = 4,
    breadcrumb = 5,
};

/// Trigger event types (ABI tags 0-5).
pub const TriggerEvent = enum(u8) {
    access = 0,
    login = 1,
    read = 2,
    write = 3,
    execute = 4,
    scan = 5,
};

/// Alert priority levels (ABI tags 0-3).
pub const AlertPriority = enum(u8) {
    low = 0,
    medium = 1,
    high = 2,
    critical = 3,
};

/// Decoy lifecycle states (ABI tags 0-3).
pub const DecoyState = enum(u8) {
    active = 0,
    triggered = 1,
    disabled = 2,
    expired = 3,
};

/// Response actions (ABI tags 0-4).
pub const ResponseAction = enum(u8) {
    alert = 0,
    redirect = 1,
    delay = 2,
    fingerprint = 3,
    isolate = 4,
};

/// Server lifecycle states (ABI tags 0-4).
pub const ServerState = enum(u8) {
    idle = 0,
    configured = 1,
    monitoring = 2,
    responding = 3,
    shutdown = 4,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent servers.
const MAX_SERVERS: usize = 64;

/// Maximum decoys per server.
const MAX_DECOYS: usize = 32;

/// Maximum alerts per server.
const MAX_ALERTS: usize = 64;

/// Maximum decoy name length in bytes.
const MAX_NAME_LEN: usize = 256;

/// A deployed decoy.
const Decoy = struct {
    /// Decoy name/identifier.
    name: [MAX_NAME_LEN]u8,
    name_len: u32,
    /// Type of decoy asset.
    decoy_type: DecoyType,
    /// Current decoy lifecycle state.
    state: DecoyState,
    /// Whether this decoy slot is active.
    active: bool,
};

/// An alert record.
const Alert = struct {
    /// Name of the decoy that triggered this alert.
    decoy_name: [MAX_NAME_LEN]u8,
    decoy_name_len: u32,
    /// The trigger event type.
    event: TriggerEvent,
    /// Alert priority.
    priority: AlertPriority,
    /// Whether this alert is still active (unresolved).
    active: bool,
};

/// A deception server instance.
const Server = struct {
    /// Current server lifecycle state.
    state: ServerState,
    /// Deployed decoys.
    decoys: [MAX_DECOYS]Decoy,
    /// Number of active decoys.
    decoy_count: u32,
    /// Alert records.
    alerts: [MAX_ALERTS]Alert,
    /// Number of active alerts.
    alert_count: u32,
    /// Total triggers handled (monotonic counter).
    triggers_handled: u64,
    /// Whether this server slot is in use.
    active: bool,
};

/// Default (empty) decoy.
const empty_decoy: Decoy = .{
    .name = [_]u8{0} ** MAX_NAME_LEN,
    .name_len = 0,
    .decoy_type = .service,
    .state = .active,
    .active = false,
};

/// Default (empty) alert.
const empty_alert: Alert = .{
    .decoy_name = [_]u8{0} ** MAX_NAME_LEN,
    .decoy_name_len = 0,
    .event = .access,
    .priority = .low,
    .active = false,
};

/// Default (empty) server.
const empty_server: Server = .{
    .state = .idle,
    .decoys = [_]Decoy{empty_decoy} ** MAX_DECOYS,
    .decoy_count = 0,
    .alerts = [_]Alert{empty_alert} ** MAX_ALERTS,
    .alert_count = 0,
    .triggers_handled = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var servers: [MAX_SERVERS]Server = [_]Server{empty_server} ** MAX_SERVERS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SERVERS) return null;
    const idx: usize = @intCast(slot);
    if (!servers[idx].active) return null;
    return idx;
}

/// Find a decoy by name within a server.
fn findDecoy(idx: usize, name: []const u8) ?usize {
    for (&servers[idx].decoys, 0..) |*d, i| {
        if (d.active and d.name_len == name.len and
            std.mem.eql(u8, d.name[0..d.name_len], name))
        {
            return i;
        }
    }
    return null;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

// -- ABI version --------------------------------------------------------------

/// Returns the ABI version number. Must match Foreign.abiVersion in Idris2.
pub export fn deception_abi_version() callconv(.c) u32 {
    return 1;
}

// -- Lifecycle ----------------------------------------------------------------

/// Create a new deception server. Returns slot index (>=0) or -1 on failure.
/// The server starts in Configured state.
pub export fn deception_create() callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    for (&servers, 0..) |*sv, i| {
        if (!sv.active) {
            sv.* = empty_server;
            sv.state = .configured;
            sv.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a server, releasing its slot.
pub export fn deception_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SERVERS) return;
    servers[@intCast(slot)] = empty_server;
}

// -- State queries ------------------------------------------------------------

/// Returns the current ServerState tag for a server.
pub export fn deception_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0; // idle fallback
    return @intFromEnum(servers[idx].state);
}

/// Returns 1 if the server can monitor, 0 otherwise.
pub export fn deception_can_monitor(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    const state = servers[idx].state;
    return if (state == .monitoring or state == .responding) 1 else 0;
}

// -- Decoy management ---------------------------------------------------------

/// Deploy a decoy. Returns 0 on success, 1 on rejection.
/// Transitions: Configured -> Monitoring, or stays Monitoring/Responding.
pub export fn deception_deploy_decoy(
    slot: c_int,
    name_ptr: [*]const u8,
    name_len: u32,
    decoy_type: u8,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = servers[idx].state;
    if (state != .configured and state != .monitoring and state != .responding) return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;
    if (decoy_type > 5) return 1;

    const name = name_ptr[0..name_len];

    // Check for duplicate decoy name
    if (findDecoy(idx, name) != null) return 1;

    // Find a free decoy slot
    for (&servers[idx].decoys) |*d| {
        if (!d.active) {
            @memcpy(d.name[0..name_len], name);
            d.name_len = name_len;
            d.decoy_type = @enumFromInt(decoy_type);
            d.state = .active;
            d.active = true;
            servers[idx].decoy_count += 1;
            if (servers[idx].state == .configured) {
                servers[idx].state = .monitoring;
            }
            return 0;
        }
    }
    return 1;
}

/// Remove a decoy by name. Returns 0 on success, 1 on rejection.
/// May transition Monitoring -> Configured if last decoy.
pub export fn deception_remove_decoy(
    slot: c_int,
    name_ptr: [*]const u8,
    name_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;

    const name = name_ptr[0..name_len];
    const di = findDecoy(idx, name) orelse return 1;

    servers[idx].decoys[di].active = false;
    servers[idx].decoy_count -= 1;

    // If no decoys remain, transition to Configured
    if (servers[idx].decoy_count == 0) {
        if (servers[idx].state == .monitoring or servers[idx].state == .responding) {
            servers[idx].state = .configured;
        }
    }

    return 0;
}

/// Returns the number of deployed decoys for a server.
pub export fn deception_decoy_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return servers[idx].decoy_count;
}

// -- Trigger / Response -------------------------------------------------------

/// Trigger a decoy alert. Returns 0 on success, 1 on rejection.
/// Transitions: Monitoring -> Responding.
pub export fn deception_trigger(
    slot: c_int,
    name_ptr: [*]const u8,
    name_len: u32,
    event: u8,
    priority: u8,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = servers[idx].state;
    if (state != .monitoring and state != .responding) return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;
    if (event > 5) return 1;
    if (priority > 3) return 1;

    const name = name_ptr[0..name_len];
    const di = findDecoy(idx, name) orelse return 1;

    if (servers[idx].decoys[di].state != .active) return 1;

    // Mark decoy as triggered
    servers[idx].decoys[di].state = .triggered;

    // Create alert
    for (&servers[idx].alerts) |*a| {
        if (!a.active) {
            @memcpy(a.decoy_name[0..name_len], name);
            a.decoy_name_len = name_len;
            a.event = @enumFromInt(event);
            a.priority = @enumFromInt(priority);
            a.active = true;
            servers[idx].alert_count += 1;
            break;
        }
    }

    servers[idx].triggers_handled += 1;
    servers[idx].state = .responding;
    return 0;
}

/// Execute a response action on a triggered decoy. Returns 0 on success, 1 on rejection.
/// May transition Responding -> Monitoring if all alerts resolved.
pub export fn deception_respond(
    slot: c_int,
    name_ptr: [*]const u8,
    name_len: u32,
    action: u8,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (servers[idx].state != .responding) return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;
    if (action > 4) return 1;

    _ = action; // Response action is recorded but not simulated

    const name = name_ptr[0..name_len];

    // Resolve the alert for this decoy
    for (&servers[idx].alerts) |*a| {
        if (a.active and a.decoy_name_len == name_len and
            std.mem.eql(u8, a.decoy_name[0..a.decoy_name_len], name))
        {
            a.active = false;
            servers[idx].alert_count -= 1;
            break;
        }
    }

    // Reset decoy state to active
    if (findDecoy(idx, name)) |di| {
        servers[idx].decoys[di].state = .active;
    }

    // If no alerts remain, return to monitoring
    if (servers[idx].alert_count == 0) {
        servers[idx].state = .monitoring;
    }

    return 0;
}

/// Returns the total number of active alerts.
pub export fn deception_alert_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return servers[idx].alert_count;
}

// -- Shutdown / Cleanup -------------------------------------------------------

/// Shutdown the server. Returns 0 on success, 1 on rejection.
pub export fn deception_shutdown(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = servers[idx].state;
    if (state == .configured or state == .monitoring or state == .responding) {
        servers[idx].state = .shutdown;
        return 0;
    }
    return 1;
}

/// Complete cleanup after shutdown. Returns 0 on success, 1 on rejection.
pub export fn deception_cleanup(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (servers[idx].state != .shutdown) return 1;

    servers[idx].state = .idle;
    servers[idx].decoys = [_]Decoy{empty_decoy} ** MAX_DECOYS;
    servers[idx].decoy_count = 0;
    servers[idx].alerts = [_]Alert{empty_alert} ** MAX_ALERTS;
    servers[idx].alert_count = 0;

    return 0;
}

// -- Stateless transition table -----------------------------------------------

/// Check if a server state transition is valid.
pub export fn deception_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Idle -> Configured
    if (from == 1 and to == 2) return 1; // Configured -> Monitoring
    if (from == 2 and to == 2) return 1; // Monitoring -> Monitoring (add more decoys)
    if (from == 2 and to == 1) return 1; // Monitoring -> Configured (all decoys removed)
    if (from == 2 and to == 3) return 1; // Monitoring -> Responding
    if (from == 3 and to == 3) return 1; // Responding -> Responding (more triggers)
    if (from == 3 and to == 2) return 1; // Responding -> Monitoring (all alerts resolved)
    if (from == 1 and to == 4) return 1; // Configured -> Shutdown
    if (from == 2 and to == 4) return 1; // Monitoring -> Shutdown
    if (from == 3 and to == 4) return 1; // Responding -> Shutdown
    if (from == 4 and to == 0) return 1; // Shutdown -> Idle
    return 0;
}

File: bindings/rust/src/deception.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//

//! Deception Platform types for the proven-servers ABI.
//!
//! Formally verified cyber deception types.
//! Mirrors the Idris2 module `DeceptionABI.Types`.
//!
//! - `DecoyType` -- Deception decoy types.
//! - `TriggerEvent` -- Decoy trigger events.
//! - `AlertPriority` -- Deception alert priority.
//! - `DecoyState` -- Decoy lifecycle states.
//! - `ResponseAction` -- Deception response actions.
//! - `ServerState` -- Deception server states.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// DecoyType (tags 0-5)
// ===========================================================================

/// Deception decoy types.
///
/// Matches `DecoyType` in `DeceptionABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum DecoyType {
    /// Service (tag 0).
    Service = 0,
    /// Credential (tag 1).
    Credential = 1,
    /// File (tag 2).
    File = 2,
    /// Network (tag 3).
    Network = 3,
    /// Token (tag 4).
    Token = 4,
    /// Breadcrumb (tag 5).
    Breadcrumb = 5,
}

impl DecoyType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Service),
            1 => Some(Self::Credential),
            2 => Some(Self::File),
            3 => Some(Self::Network),
            4 => Some(Self::Token),
            5 => Some(Self::Breadcrumb),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [DecoyType; 6] = [
        Self::Service, Self::Credential, Self::File, Self::Network, Self::Token, Self::Breadcrumb,
    ];
}

impl fmt::Display for DecoyType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// TriggerEvent (tags 0-5)
// ===========================================================================

/// Decoy trigger events.
///
/// Matches `TriggerEvent` in `DeceptionABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum TriggerEvent {
    /// Access (tag 0).
    Access = 0,
    /// Login (tag 1).
    Login = 1,
    /// Read (tag 2).
    Read = 2,
    /// Write (tag 3).
    Write = 3,
    /// Execute (tag 4).
    Execute = 4,
    /// Scan (tag 5).
    Scan = 5,
}

impl TriggerEvent {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Access),
            1 => Some(Self::Login),
            2 => Some(Self::Read),
            3 => Some(Self::Write),
            4 => Some(Self::Execute),
            5 => Some(Self::Scan),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [TriggerEvent; 6] = [
        Self::Access, Self::Login, Self::Read, Self::Write, Self::Execute, Self::Scan,
    ];
}

impl fmt::Display for TriggerEvent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// AlertPriority (tags 0-3)
// ===========================================================================

/// Deception alert priority.
///
/// Matches `AlertPriority` in `DeceptionABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum AlertPriority {
    /// Low (tag 0).
    Low = 0,
    /// Medium (tag 1).
    Medium = 1,
    /// High (tag 2).
    High = 2,
    /// Critical (tag 3).
    Critical = 3,
}

impl AlertPriority {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Low),
            1 => Some(Self::Medium),
            2 => Some(Self::High),
            3 => Some(Self::Critical),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [AlertPriority; 4] = [
        Self::Low, Self::Medium, Self::High, Self::Critical,
    ];
}

impl fmt::Display for AlertPriority {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// DecoyState (tags 0-3)
// ===========================================================================

/// Decoy lifecycle states.
///
/// Matches `DecoyState` in `DeceptionABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum DecoyState {
    /// Active (tag 0).
    Active = 0,
    /// Triggered (tag 1).
    Triggered = 1,
    /// Disabled (tag 2).
    Disabled = 2,
    /// Expired (tag 3).
    Expired = 3,
}

impl DecoyState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Active),
            1 => Some(Self::Triggered),
            2 => Some(Self::Disabled),
            3 => Some(Self::Expired),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [DecoyState; 4] = [
        Self::Active, Self::Triggered, Self::Disabled, Self::Expired,
    ];
}

impl fmt::Display for DecoyState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// ResponseAction (tags 0-4)
// ===========================================================================

/// Deception response actions.
///
/// Matches `ResponseAction` in `DeceptionABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum ResponseAction {
    /// Alert (tag 0).
    Alert = 0,
    /// Redirect (tag 1).
    Redirect = 1,
    /// Delay (tag 2).
    Delay = 2,
    /// Fingerprint (tag 3).
    Fingerprint = 3,
    /// Isolate (tag 4).
    Isolate = 4,
}

impl ResponseAction {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Alert),
            1 => Some(Self::Redirect),
            2 => Some(Self::Delay),
            3 => Some(Self::Fingerprint),
            4 => Some(Self::Isolate),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [ResponseAction; 5] = [
        Self::Alert, Self::Redirect, Self::Delay, Self::Fingerprint, Self::Isolate,
    ];
}

impl fmt::Display for ResponseAction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// ServerState (tags 0-4)
// ===========================================================================

/// Deception server states.
///
/// Matches `ServerState` in `DeceptionABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum ServerState {
    /// Idle (tag 0).
    Idle = 0,
    /// Configured (tag 1).
    Configured = 1,
    /// Monitoring (tag 2).
    Monitoring = 2,
    /// Responding (tag 3).
    Responding = 3,
    /// Shutdown (tag 4).
    Shutdown = 4,
}

impl ServerState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Idle),
            1 => Some(Self::Configured),
            2 => Some(Self::Monitoring),
            3 => Some(Self::Responding),
            4 => Some(Self::Shutdown),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [ServerState; 5] = [
        Self::Idle, Self::Configured, Self::Monitoring, Self::Responding, Self::Shutdown,
    ];
}

impl fmt::Display for ServerState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn decoy_type_roundtrip() {
        for v in DecoyType::ALL {
            let tag = v.to_tag();
            let decoded = DecoyType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(DecoyType::from_tag(6).is_none());
    }

    #[test]
    fn trigger_event_roundtrip() {
        for v in TriggerEvent::ALL {
            let tag = v.to_tag();
            let decoded = TriggerEvent::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(TriggerEvent::from_tag(6).is_none());
    }

    #[test]
    fn alert_priority_roundtrip() {
        for v in AlertPriority::ALL {
            let tag = v.to_tag();
            let decoded = AlertPriority::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(AlertPriority::from_tag(4).is_none());
    }

    #[test]
    fn decoy_state_roundtrip() {
        for v in DecoyState::ALL {
            let tag = v.to_tag();
            let decoded = DecoyState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(DecoyState::from_tag(4).is_none());
    }

    #[test]
    fn response_action_roundtrip() {
        for v in ResponseAction::ALL {
            let tag = v.to_tag();
            let decoded = ResponseAction::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(ResponseAction::from_tag(5).is_none());
    }

    #[test]
    fn server_state_roundtrip() {
        for v in ServerState::ALL {
            let tag = v.to_tag();
            let decoded = ServerState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(ServerState::from_tag(5).is_none());
    }

}
